The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/26406360004/job/77730606646) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/77730606646) [(plain)](https://github.com/rust-lang/rust/commit/68b6a99c4dda376c2a8e8a5d4fa6f6c5ac009790/checks/77730606646/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt: checked 6864 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
tidy [style (tests)]: /checkout/tests/ui/eii/unused_extern_crate.rs:3: line longer than 100 chars
tidy [style (tests)]: FAIL
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (26.1.1)
linting python files
---
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: style (tests)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1618:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1417:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1417:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1595:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1339:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1138:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1117:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:49
  local time: Mon May 25 14:55:06 UTC 2026
  network time: Mon, 25 May 2026 14:55:06 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/compiler/rustc_passes/src/eii.rs
+++ b/compiler/rustc_passes/src/eii.rs
@@ -141,6 +141,7 @@ pub(crate) fn check_externally_implementable_items<'tcx>(tcx: TyCtxt<'tcx>, (): 
                         decl_crate_name: tcx.crate_name(decl_crate),
                         // FIXME: shouldn't call `item_name`
                         name: decl.name.name,
+                        kind: tcx.def_kind(decl.foreign_item).descr(decl.foreign_item),
                         span: decl.name.span,
                         help: (),
                     });

--- a/compiler/rustc_passes/src/errors.rs
+++ b/compiler/rustc_passes/src/errors.rs
@@ -1209,12 +1209,13 @@ pub(crate) struct EiiWithTrackCaller {
 }
 
 #[derive(Diagnostic)]
-#[diag("`#[{$name}]` required, but not found")]
+#[diag("`#[{$name}]` {$kind} required, but not found")]
 pub(crate) struct EiiWithoutImpl {
     #[primary_span]
     #[label("expected because `#[{$name}]` was declared here in crate `{$decl_crate_name}`")]
     pub span: Span,
     pub name: Symbol,
+    pub kind: &'static str,
 
     pub current_crate_name: Symbol,
     pub decl_crate_name: Symbol,

--- a/compiler/rustc_resolve/src/check_unused.rs
+++ b/compiler/rustc_resolve/src/check_unused.rs
@@ -438,7 +438,10 @@ impl Resolver<'_, '_> {
                             && !tcx.is_panic_runtime(cnum)
                             && !tcx.has_global_allocator(cnum)
                             && !tcx.has_panic_handler(cnum)
-                            && tcx.externally_implementable_items(cnum).is_empty()
+                            && tcx
+                                .externally_implementable_items(cnum)
+                                .values()
+                                .all(|(_, defs)| defs.is_empty())
                     }) {
                         maybe_unused_extern_crates.insert(id, import.span);
                     }

--- a/tests/ui/eii/auxiliary/unused_extern_crate_decl.rs
+++ b/tests/ui/eii/auxiliary/unused_extern_crate_decl.rs
@@ -1,0 +1,6 @@
+//@ no-prefer-dynamic
+#![crate_type = "rlib"]
+#![feature(extern_item_impls)]
+
+#[eii(eii1)]
+pub fn decl1(x: u64);

--- a/tests/ui/eii/auxiliary/unused_extern_crate_impl.rs
+++ b/tests/ui/eii/auxiliary/unused_extern_crate_impl.rs
@@ -1,0 +1,9 @@
+//@ no-prefer-dynamic
+//@ aux-build:unused_extern_crate_decl.rs
+#![crate_type = "rlib"]
+#![feature(extern_item_impls)]
+
+extern crate unused_extern_crate_decl;
+
+#[unused_extern_crate_decl::eii1]
+fn impl1(x: u64) {}

--- a/tests/ui/eii/dylib_needs_impl.rs
+++ b/tests/ui/eii/dylib_needs_impl.rs
@@ -3,5 +3,5 @@
 #![crate_type = "dylib"]
 #![feature(extern_item_impls)]
 
-#[eii(eii1)] //~ ERROR `#[eii1]` required, but not found
+#[eii(eii1)] //~ ERROR `#[eii1]` function required, but not found
 fn decl1(x: u64);

--- a/tests/ui/eii/dylib_needs_impl.stderr
+++ b/tests/ui/eii/dylib_needs_impl.stderr
@@ -1,4 +1,4 @@
-error: `#[eii1]` required, but not found
+error: `#[eii1]` function required, but not found
   --> $DIR/dylib_needs_impl.rs:6:7
    |
 LL | #[eii(eii1)]

--- a/tests/ui/eii/privacy2.rs
+++ b/tests/ui/eii/privacy2.rs
@@ -5,8 +5,8 @@
 extern crate other_crate_privacy2 as codegen;
 
 // has a span but in the other crate
-//~? ERROR `#[eii2]` required, but not found
-//~? ERROR `#[eii3]` required, but not found
+//~? ERROR `#[eii2]` function required, but not found
+//~? ERROR `#[eii3]` function required, but not found
 
 #[codegen::eii1]
 fn eii1_impl(x: u64) {

--- a/tests/ui/eii/privacy2.stderr
+++ b/tests/ui/eii/privacy2.stderr
@@ -16,7 +16,7 @@ note: the function `decl1` is defined here
 LL | fn decl1(x: u64);
    | ^^^^^^^^^^^^^^^^^
 
-error: `#[eii2]` required, but not found
+error: `#[eii2]` function required, but not found
   --> $DIR/auxiliary/other_crate_privacy2.rs:9:7
    |
 LL | #[eii(eii2)]
@@ -24,7 +24,7 @@ LL | #[eii(eii2)]
    |
    = help: expected at least one implementation in crate `privacy2` or any of its dependencies
 
-error: `#[eii3]` required, but not found
+error: `#[eii3]` function required, but not found
   --> $DIR/auxiliary/other_crate_privacy2.rs:13:11
    |
 LL |     #[eii(eii3)]

--- a/tests/ui/eii/shadow_builtin.rs
+++ b/tests/ui/eii/shadow_builtin.rs
@@ -5,7 +5,7 @@
 #![feature(extern_item_impls)]
 
 #[eii(inline)]
-//~^ ERROR `#[inline]` required, but not found
+//~^ ERROR `#[inline]` function required, but not found
 fn test(x: u64);
 
 #[inline]
@@ -14,4 +14,4 @@ fn test_impl(x: u64) {
     println!("{x:?}")
 }
 
-fn main() { }
+fn main() {}

--- a/tests/ui/eii/shadow_builtin.stderr
+++ b/tests/ui/eii/shadow_builtin.stderr
@@ -13,7 +13,7 @@ LL | #[eii(inline)]
    | ^^^^^^^^^^^^^^
    = help: use `crate::inline` to refer to this attribute macro unambiguously
 
-error: `#[inline]` required, but not found
+error: `#[inline]` function required, but not found
   --> $DIR/shadow_builtin.rs:7:7
    |
 LL | #[eii(inline)]

--- a/tests/ui/eii/unused_extern_crate.rs
+++ b/tests/ui/eii/unused_extern_crate.rs
@@ -1,0 +1,11 @@
+//@ aux-build:unused_extern_crate_decl.rs
+//@ aux-build:unused_extern_crate_impl.rs
+// Tests that dependencies that contain an EII decl without any EII impl are
+// still considered unused.
+#![feature(extern_item_impls)]
+#![deny(unused_extern_crates)]
+
+extern crate unused_extern_crate_decl; //~ ERROR unused extern crate
+extern crate unused_extern_crate_impl;
+
+fn main() {}

--- a/tests/ui/eii/unused_extern_crate.stderr
+++ b/tests/ui/eii/unused_extern_crate.stderr
@@ -1,0 +1,19 @@
+error: unused extern crate
+  --> $DIR/unused_extern_crate.rs:8:1
+   |
+LL | extern crate unused_extern_crate_decl;
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unused
+   |
+note: the lint level is defined here
+  --> $DIR/unused_extern_crate.rs:6:9
+   |
+LL | #![deny(unused_extern_crates)]
+   |         ^^^^^^^^^^^^^^^^^^^^
+help: remove the unused `extern crate`
+   |
+LL - extern crate unused_extern_crate_decl;
+LL +
+   |
+
+error: aborting due to 1 previous error
+

